compiler_opt/es/blackbox_learner.py

```diff
@@ -0,0 +1,314 @@
+
+  def _update_model(self, perturbations: List[npt.NDArray[np.float32]],
+                    rewards: List[float]) -> None:
+    """Update the model given a list of perturbations and rewards."""
+    self._model_weights = self._blackbox_opt.run_step(
+        perturbations=np.array(perturbations),
+        function_values=np.array(rewards),
+        current_input=self._model_weights,
+        current_value=np.mean(rewards))
+
+  def _log_rewards(self, rewards: List[float]) -> None:
+    """Log reward to console."""
+    logging.info('Train reward: [%f]', np.mean(rewards))
+
+  def _log_tf_summary(self, rewards: List[float]) -> None:
+    """Log tensorboard data."""
+    with self._summary_writer.as_default():
+      tf.summary.scalar(
+          'reward/average_reward_train', np.mean(rewards), step=self._step)
+
+      tf.summary.histogram('reward/reward_train', rewards, step=self._step)
+
+      train_regressions = [reward for reward in rewards if reward < 0]
+      tf.summary.scalar(
+          'reward/regression_probability_train',
+          len(train_regressions) / len(rewards),
+          step=self._step)
+
+      tf.summary.scalar(
+          'reward/regression_avg_train',
+          np.mean(train_regressions),
+          step=self._step)
+
+      # The "max regression" is the min value, as the regressions are negative.
+      tf.summary.scalar(
+          'reward/regression_max_train',
+          min(train_regressions, default=0),
+          step=self._step)
+
+      train_wins = [reward for reward in rewards if reward > 0]
+      tf.summary.scalar(
+          'reward/win_probability_train',
+          len(train_wins) / len(rewards),
+          step=self._step)
+
+  def _save_model(self) -> None:
+    """Save the model."""
+    logging.info('Saving the model.')
+    self._policy_saver_fn(
+        parameters=self._model_weights, policy_name=f'iteration{self._step}')
+
+  def get_model_weights(self) -> npt.NDArray[np.float32]:
+    return self._model_weights
+
+  def _get_results(
+      self, pool: FixedWorkerPool,
+      perturbations: List[bytes]) -> List[concurrent.futures.Future]:
+    if not self._samples:
+      for _ in range(self._config.total_num_perturbations):
+        sample = self._sampler.sample(self._config.num_ir_repeats_within_worker)
+        self._samples.append(sample)
+        # add copy of sample for antithetic perturbation pair
+        if self._config.est_type == (
+            blackbox_optimizers.EstimatorType.ANTITHETIC):
+          self._samples.append(sample)
+
+    compile_args = zip(perturbations, self._samples)
+
+    _, futures = buffered_scheduler.schedule_on_worker_pool(
+        action=lambda w, v: w.temp_compile(v[0], v[1]),
+        jobs=compile_args,
+        worker_pool=pool)
+
+    not_done = futures
+    # wait for all futures to finish
+    while not_done:
+      # update lists as work gets done
+      _, not_done = concurrent.futures.wait(
+          not_done, return_when='FIRST_COMPLETED')
```

Comment:
I think there's a constant `concurrent.futures.FIRST_COMPLETED` - so we avoid strings.